--- resnet_utils.py	2024-06-05 11:09:37.000000000 +0900
+++ ../resnet_utils_2.py	2024-05-22 18:54:50.000000000 +0900
@@ -13,20 +13,16 @@
 # limitations under the License.
 # ==============================================================================
 """Contains building blocks for various versions of Residual Networks.
-
 Residual networks (ResNets) were proposed in:
   Kaiming He, Xiangyu Zhang, Shaoqing Ren, Jian Sun
   Deep Residual Learning for Image Recognition. arXiv:1512.03385, 2015
-
 More variants were introduced in:
   Kaiming He, Xiangyu Zhang, Shaoqing Ren, Jian Sun
   Identity Mappings in Deep Residual Networks. arXiv: 1603.05027, 2016
-
 We can obtain different ResNet variants by changing the network depth, width,
 and form of residual unit. This module implements the infrastructure for
 building them. Concrete ResNet units and full ResNet networks are implemented in
 the accompanying resnet_v1.py and resnet_v2.py modules.
-
 Compared to https://github.com/KaimingHe/deep-residual-networks, in the current
 implementation we subsample the output activations in the last residual unit of
 each block, instead of subsampling the input activations in the first residual
@@ -38,13 +34,11 @@
 from __future__ import print_function
 
 import collections
-import tensorflow.compat.v1 as tf
-import tf_slim as slim
-
-
-class Block(collections.namedtuple('Block', ['scope', 'unit_fn', 'args'])):
+import tensorflow as tf
+import numpy as np
+slim = tf.contrib.slim
+class Block(collections.namedtuple('Block', ['scope','unit_fn', 'args'])):
   """A named tuple describing a ResNet block.
-
   Its parts are:
     scope: The scope of the `Block`.
     unit_fn: The ResNet unit function which takes as input a `Tensor` and
@@ -57,12 +51,10 @@
 
 def subsample(inputs, factor, scope=None):
   """Subsamples the input along the spatial dimensions.
-
   Args:
     inputs: A `Tensor` of size [batch, height_in, width_in, channels].
     factor: The subsampling factor.
     scope: Optional variable_scope.
-
   Returns:
     output: A `Tensor` of size [batch, height_out, width_out, channels] with the
       input, either intact (if factor == 1) or subsampled (if factor > 1).
@@ -72,29 +64,37 @@
   else:
     return slim.max_pool2d(inputs, [1, 1], stride=factor, scope=scope)
 
+def get_p_survival(block=1, nb_total_blocks=50, p_survival_end=0.5, mode='linear_decay'):
+    """
+    See eq. (4) in stochastic depth paper: http://arxiv.org/pdf/1603.09382v1.pdf
+    """
+    if mode == 'uniform':
+        return p_survival_end
+    elif mode == 'linear_decay':
+        return 1 - (block / nb_total_blocks) * (1 - p_survival_end)
+
+def stochastic_survival(p_survival=1.0):
+    # binomial random variable
+    survival = np.random.binomial(1, p=p_survival)
+    # during testing phase:
+    # - scale y (see eq. (6))
+    # - p_survival effectively becomes 1 for all layers (no layer dropout)
+    return survival
+
 
 def conv2d_same(inputs, num_outputs, kernel_size, stride, rate=1, scope=None):
   """Strided 2-D convolution with 'SAME' padding.
-
   When stride > 1, then we do explicit zero-padding, followed by conv2d with
   'VALID' padding.
-
   Note that
-
      net = conv2d_same(inputs, num_outputs, 3, stride=stride)
-
   is equivalent to
-
      net = slim.conv2d(inputs, num_outputs, 3, stride=1, padding='SAME')
      net = subsample(net, factor=stride)
-
   whereas
-
      net = slim.conv2d(inputs, num_outputs, 3, stride=stride, padding='SAME')
-
   is different when the input's height or width is even, which is why we add the
   current function. For more details, see ResnetUtilsTest.testConv2DSameEven().
-
   Args:
     inputs: A 4-D tensor of size [batch, height_in, width_in, channels].
     num_outputs: An integer, the number of output filters.
@@ -102,7 +102,6 @@
     stride: An integer, the output stride.
     rate: An integer, rate for atrous convolution.
     scope: Scope.
-
   Returns:
     output: A 4-D tensor of size [batch, height_out, width_out, channels] with
       the convolution output.
@@ -115,35 +114,29 @@
     pad_total = kernel_size_effective - 1
     pad_beg = pad_total // 2
     pad_end = pad_total - pad_beg
-    inputs = tf.pad(
-        tensor=inputs,
-        paddings=[[0, 0], [pad_beg, pad_end], [pad_beg, pad_end], [0, 0]])
+    inputs = tf.pad(inputs,
+                    [[0, 0], [pad_beg, pad_end], [pad_beg, pad_end], [0, 0]])
     return slim.conv2d(inputs, num_outputs, kernel_size, stride=stride,
                        rate=rate, padding='VALID', scope=scope)
 
 
 @slim.add_arg_scope
-def stack_blocks_dense(net, blocks, output_stride=None,
+def stack_blocks_dense(net,y, blocks, output_stride=None,
                        store_non_strided_activations=False,
                        outputs_collections=None):
   """Stacks ResNet `Blocks` and controls output feature density.
-
   First, this function creates scopes for the ResNet in the form of
   'block_name/unit_1', 'block_name/unit_2', etc.
-
   Second, this function allows the user to explicitly control the ResNet
   output_stride, which is the ratio of the input to output spatial resolution.
   This is useful for dense prediction tasks such as semantic segmentation or
   object detection.
-
   Most ResNets consist of 4 ResNet blocks and subsample the activations by a
   factor of 2 when transitioning between consecutive ResNet blocks. This results
   to a nominal ResNet output_stride equal to 8. If we set the output_stride to
   half the nominal network stride (e.g., output_stride=4), then we compute
   responses twice.
-
   Control of the output feature density is implemented by atrous convolution.
-
   Args:
     net: A `Tensor` of size [batch, height, width, channels].
     blocks: A list of length equal to the number of ResNet `Blocks`. Each
@@ -162,10 +155,8 @@
       dense prediction problems but increases 4x the computation and memory cost
       at the last unit of each block.
     outputs_collections: Collection to add the ResNet block outputs.
-
   Returns:
     net: Output tensor with stride equal to the specified output_stride.
-
   Raises:
     ValueError: If the target output_stride is not valid.
   """
@@ -177,11 +168,13 @@
 
   # The atrous convolution rate parameter.
   rate = 1
-
+  k=0
   for block in blocks:
     with tf.variable_scope(block.scope, 'block', [net]) as sc:
       block_stride = 1
       for i, unit in enumerate(block.args):
+        #p_survival = get_p_survival(block=k, nb_total_blocks=50, p_survival_end=0.5, mode='linear_decay')
+        #p= np.random.binomial(1, p=p_survival)
         if store_non_strided_activations and i == len(block.args) - 1:
           # Move stride from the block's last unit to the end of the block.
           block_stride = unit.get('stride', 1)
@@ -192,18 +185,16 @@
           # atrous convolution with stride=1 and multiply the atrous rate by the
           # current unit's stride for use in subsequent layers.
           if output_stride is not None and current_stride == output_stride:
-            net = block.unit_fn(net, rate=rate, **dict(unit, stride=1))
+            net = block.unit_fn(net,rate=rate, **dict(unit, stride=1),p = y[k])
             rate *= unit.get('stride', 1)
-
           else:
-            net = block.unit_fn(net, rate=1, **unit)
+            net = block.unit_fn(net,rate=1, **unit, p = y[k])
             current_stride *= unit.get('stride', 1)
             if output_stride is not None and current_stride > output_stride:
               raise ValueError('The target output_stride cannot be reached.')
-
+          k=k+1
       # Collect activations at the block's end before performing subsampling.
       net = slim.utils.collect_named_outputs(outputs_collections, sc.name, net)
-
       # Subsampling of the block's output activations.
       if output_stride is not None and current_stride == output_stride:
         rate *= block_stride
@@ -215,25 +206,21 @@
 
   if output_stride is not None and current_stride != output_stride:
     raise ValueError('The target output_stride cannot be reached.')
-
   return net
 
 
-def resnet_arg_scope(
-    weight_decay=0.0001,
-    batch_norm_decay=0.997,
-    batch_norm_epsilon=1e-5,
-    batch_norm_scale=True,
-    activation_fn=tf.nn.relu,
-    use_batch_norm=True,
-    batch_norm_updates_collections=tf.GraphKeys.UPDATE_OPS):
+def resnet_arg_scope(weight_decay=0.0001,
+                     batch_norm_decay=0.997,
+                     batch_norm_epsilon=1e-5,
+                     batch_norm_scale=True,
+                     activation_fn=tf.nn.relu,
+                     use_batch_norm=True,
+                     batch_norm_updates_collections=tf.GraphKeys.UPDATE_OPS):
   """Defines the default ResNet arg scope.
-
   TODO(gpapan): The batch-normalization related default values above are
     appropriate for use in conjunction with the reference ResNet models
     released at https://github.com/KaimingHe/deep-residual-networks. When
     training ResNets from scratch, they might need to be tuned.
-
   Args:
     weight_decay: The weight decay to use for regularizing the model.
     batch_norm_decay: The moving average decay when estimating layer activation
@@ -246,7 +233,6 @@
     use_batch_norm: Whether or not to use batch normalization.
     batch_norm_updates_collections: Collection for the update ops for
       batch norm.
-
   Returns:
     An `arg_scope` to use for the resnet models.
   """
