--- resnet_v2.py	2024-06-05 11:09:37.000000000 +0900
+++ ../resnet_v2_2.py	2024-06-05 09:09:00.000000000 +0900
@@ -13,31 +13,22 @@
 # limitations under the License.
 # ==============================================================================
 """Contains definitions for the preactivation form of Residual Networks.
-
 Residual networks (ResNets) were originally proposed in:
 [1] Kaiming He, Xiangyu Zhang, Shaoqing Ren, Jian Sun
     Deep Residual Learning for Image Recognition. arXiv:1512.03385
-
 The full preactivation 'v2' ResNet variant implemented in this module was
 introduced by:
 [2] Kaiming He, Xiangyu Zhang, Shaoqing Ren, Jian Sun
     Identity Mappings in Deep Residual Networks. arXiv: 1603.05027
-
 The key difference of the full preactivation 'v2' variant compared to the
 'v1' variant in [1] is the use of batch normalization before every weight layer.
-
 Typical use:
-
-   from tf_slim.nets import resnet_v2
-
+   from tensorflow.contrib.slim.nets import resnet_v2
 ResNet-101 for image classification into 1000 classes:
-
    # inputs has shape [batch, 224, 224, 3]
    with slim.arg_scope(resnet_v2.resnet_arg_scope()):
       net, end_points = resnet_v2.resnet_v2_101(inputs, 1000, is_training=False)
-
 ResNet-101 for semantic segmentation into 21 classes:
-
    # inputs has shape [batch, 513, 513, 3]
    with slim.arg_scope(resnet_v2.resnet_arg_scope()):
       net, end_points = resnet_v2.resnet_v2_101(inputs,
@@ -50,11 +41,11 @@
 from __future__ import division
 from __future__ import print_function
 
-import tensorflow.compat.v1 as tf
-import tf_slim as slim
+import tensorflow as tf
 
-from nets import resnet_utils
+import resnet_utils
 
+slim = tf.contrib.slim
 resnet_arg_scope = resnet_utils.resnet_arg_scope
 
 
@@ -62,14 +53,11 @@
 def bottleneck(inputs, depth, depth_bottleneck, stride, rate=1,
                outputs_collections=None, scope=None):
   """Bottleneck residual unit variant with BN before convolutions.
-
   This is the full preactivation residual unit variant proposed in [2]. See
   Fig. 1(b) of [2] for its definition. Note that we use here the bottleneck
   variant which has an extra bottleneck layer.
-
   When putting together two consecutive ResNet blocks that use this unit, one
   should use stride = 2 in the last unit of the first block.
-
   Args:
     inputs: A tensor of size [batch, height, width, channels].
     depth: The depth of the ResNet unit output.
@@ -79,7 +67,6 @@
     rate: An integer, rate for atrous convolution.
     outputs_collections: Collection to add the ResNet unit output.
     scope: Optional variable_scope.
-
   Returns:
     The ResNet unit's output.
   """
@@ -119,11 +106,9 @@
               reuse=None,
               scope=None):
   """Generator for v2 (preactivation) ResNet models.
-
   This function generates a family of ResNet v2 models. See the resnet_v2_*()
   methods for specific model instantiations, obtained by selecting different
   block instantiations that produce ResNets of various depths.
-
   Training for image classification on Imagenet is usually done with [224, 224]
   inputs, resulting in [7, 7] feature maps at the output of the last ResNet
   block for the ResNets defined in [1] that have nominal stride equal to 32.
@@ -134,13 +119,11 @@
   and corners exactly aligned with the input image corners, which greatly
   facilitates alignment of the features to the image. Using as input [225, 225]
   images results in [8, 8] feature maps at the output of the last ResNet block.
-
   For dense prediction tasks, the ResNet needs to run in fully-convolutional
   (FCN) mode and global_pool needs to be set to False. The ResNets in [1, 2] all
   have nominal stride equal to 32 and a good choice in FCN mode is to use
   output_stride=16 in order to increase the density of the computed features at
   small computational and memory overhead, cf. http://arxiv.org/abs/1606.00915.
-
   Args:
     inputs: A tensor of size [batch, height_in, width_in, channels].
     blocks: A list of length equal to the number of ResNet blocks. Each element
@@ -164,8 +147,6 @@
     reuse: whether or not the network and its variables should be reused. To be
       able to reuse 'scope' must be given.
     scope: Optional variable_scope.
-
-
   Returns:
     net: A rank-4 tensor of size [batch, height_out, width_out, channels_out].
       If global_pool is False, then height_out and width_out are reduced by a
@@ -176,12 +157,10 @@
       pre-softmax activations.
     end_points: A dictionary from components of the network to the corresponding
       activation.
-
   Raises:
     ValueError: If the target output_stride is not valid.
   """
-  with tf.variable_scope(
-      scope, 'resnet_v2', [inputs], reuse=reuse) as sc:
+  with tf.variable_scope(scope, 'resnet_v2', [inputs], reuse=reuse) as sc:
     end_points_collection = sc.original_name_scope + '_end_points'
     with slim.arg_scope([slim.conv2d, bottleneck,
                          resnet_utils.stack_blocks_dense],
@@ -211,8 +190,8 @@
 
         if global_pool:
           # Global average pooling.
-          net = tf.reduce_mean(
-              input_tensor=net, axis=[1, 2], name='pool5', keepdims=True)
+         # net = [tf.reduce_mean(net[i:i+49], [0, 1, 2], name='pool5', keep_dims=True) for i in range(0,147,49)]
+          net = tf.reduce_mean(net, [0, 1, 2], name='pool5', keep_dims=True)
           end_points['global_pool'] = net
         if num_classes:
           net = slim.conv2d(net, num_classes, [1, 1], activation_fn=None,
@@ -228,14 +207,12 @@
 
 def resnet_v2_block(scope, base_depth, num_units, stride):
   """Helper function for creating a resnet_v2 bottleneck block.
-
   Args:
     scope: The scope of the block.
     base_depth: The depth of the bottleneck layer for each unit.
     num_units: The number of units in the block.
     stride: The stride of the block, implemented as a stride in the last unit.
       All other units have stride=1.
-
   Returns:
     A resnet_v2 bottleneck block.
   """
